src/client/pythonEnvironments/discovery/locators/services/pipEnvHelper.ts

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import * as path from 'path';
+import { getEnvironmentVariable } from '../../../../common/utils/platform';
+import { pathExists, readFile } from '../../../common/externalDependencies';
+
+/**
+ * Returns the path to Pipfile associated with the provided directory.
+ * @param searchDir the directory to look into
+ * @param lookIntoParentDirectories set to true if we should also search for Pipfile in parent directory
+ */
+async function getAssociatedPipfile(
+    searchDir: string,
+    lookIntoParentDirectories: boolean,
+): Promise<string | undefined> {
+    const pipFileName = getEnvironmentVariable('PIPENV_PIPFILE') || 'Pipfile';
+    let depthToSearch = 1;
+    if (lookIntoParentDirectories) {
+        // PIPENV_MAX_DEPTH tells pipenv the maximum number of directories to recursively search for
+        // a Pipfile, defaults to 3: https://pipenv.pypa.io/en/latest/advanced/#pipenv.environments.PIPENV_MAX_DEPTH
+        const maxDepth = getEnvironmentVariable('PIPENV_MAX_DEPTH');
+        if (maxDepth) {
+            depthToSearch = +maxDepth;
+        } else {
+            depthToSearch = 3;
+        }
+    }
+    while (depthToSearch > 0 && searchDir !== path.dirname(searchDir)) {

```

Comment:
Does `path.dirname()` handle OS-specific path separators properly?  If not, `searchDir !== path.dirname(searchDir)` may be problematic.  In that case, be sure to normalize first.

See:
* https://nodejs.org/api/path.html#path_windows_vs_posix
* `FileSystemPaths.normCase()` in `src/client/common/platform/fs-paths.ts` (though that isn't as reliable as the first link)

```suggestion
    searchDir = path.
    while (depthToSearch > 0 && searchDir !== path.dirname(searchDir)) {
```

Reply:
> Does path.dirname() handle OS-specific path separators properly?

Yes, it yields different results for different OS, taking path separators into account. I'll still normalize to be safe.